src/lib/SaferNames/Name.hs

```diff
@@ -0,0 +1,478 @@
+
+freeNames :: HasNamesE e => e n -> NameSet n
+freeNames e = foldMapNames S.singleton e
+
+-- should be equivalent to `\ext -> fmapNames (injectName ext)`
+injectNamesL :: HasNamesE e => FreshExt n l -> e n -> e l
+injectNamesL _ = unsafeCoerceE
+
+projectNamesL :: HasNamesE e => Scope n -> Scope (n:=>:l) -> e l -> Maybe (e n)
+projectNamesL scope scopeFragment e = traverseNamesE scope t e
+  where t = newNameTraversal \name -> case projectName scopeFragment name of
+                                        Left name' -> Just name'
+                                        Right _    -> Nothing
+
+-- === common scoping patterns ===
+
+data Abs (binder::B) (body::E) (n::S) where
+  Abs :: binder n l -> body l -> Abs binder body n
+
+data FreshAbs (binder::B) (body::E) (n::S) where
+  FreshAbs :: FreshExt n l -> binder n l -> body l -> FreshAbs binder body n
+
+data NestPair (b1::B) (b2::B) (n::S) (l::S) where
+  NestPair :: b1 n l -> b2 l l' -> NestPair b1 b2 n l'
+
+data Nest (binder::B) (n::S) (l::S) where
+  Nest  :: binder n h -> Nest binder h l -> Nest binder n l
+  Empty ::                                  Nest binder n n
+
+type PlainBinderList = Nest PlainBinder
+
+infixl 7 :>
+data AnnBinderP (b::B) (ann ::E) (n::S) (l::S) =
+  (:>) (b n l) (ann n)
+  deriving (Show)
+
+infixl 7 :>>
+data AnnBinderListP (b::B) (ann::E) (n::S) (l::S) =
+  (:>>) (Nest b n l) [ann n]
+
+-- === environment variants ===
+
+type RenameEnv i o = Env i (Name o)
+
+-- covariant in `o`, contravariant in `i`.
+newtype EnvE (e::E) (i::S) (o::S) = EnvE { fromEnvE :: Env i (e o) }
+
+-- covariant in `n`
+newtype RecEnv (e::E) (n::S) = RecEnv { fromRecEnv :: Env n (e n) }
+
+-- covariant in `n`, contravariant in `l`
+newtype RecEnvFrag (e::E) (n::S) (l::S) = RecEnvFrag { fromRecEnvFrag :: Env (n:=>:l) (e l)}
+
+-- === various E-kind and B-kind versions of standard containers and classes ===
+
+type PrettyE  e = (forall (n::S)       . Pretty (e n  )) :: Constraint
+type PrettyB b  = (forall (n::S) (l::S). Pretty (b n l)) :: Constraint
+
+type ShowE e = (forall (n::S)       . Show (e n  )) :: Constraint
+type ShowB b = (forall (n::S) (l::S). Show (b n l)) :: Constraint
+
+type EqE e = (forall (n::S)       . Eq (e n  )) :: Constraint
+type EqB b = (forall (n::S) (l::S). Eq (b n l)) :: Constraint
+
+data UnitE (n::S) = UnitH deriving (Show, Eq)
+data VoidE (n::S)
+
+data PairE (e1::E) (e2::E) (n::S) = PairE (e1 n) (e2 n)
+     deriving (Show, Eq)
+
+data EitherE (e1::E) (e2::E) (n::S) = LeftE (e1 n) | RightE (e2 n)
+     deriving (Show, Eq)
+
+data MaybeE (e::E) (n::S) = JustE (e n) | NothingE
+     deriving (Show, Eq)
+
+data ListE (e::E) (n::S) = ListE { fromListE :: [e n] }
+     deriving (Show, Eq)
+
+data LiftE (a:: *) (n::S) = LiftE { fromLiftE :: a }
+     deriving (Show, Eq)
+
+-- === various convenience utilities ===
+
+class HasNamesE e => AlphaEq (e::E) where
+  alphaEq :: Scope n -> e n -> e n -> Bool
```

Comment:
Not yet. I tried to separate the changes cleanly but there are some fragments like these that refer to things that get used in follow-up changes. This will be used in the type checking pass which is the next PR.